JSONAPI/Core/ModelManager.cs

```diff
@@ -0,0 +1,137 @@
+        {
+            PropertyInfo idprop = null;
+
+            var idPropCache = _idProperties.Value;
+
+            lock (idPropCache)
+            {
+                if (idPropCache.TryGetValue(type, out idprop)) return idprop;
+
+                //TODO: Enable attribute-based determination
+
+                idprop = type.GetProperty("Id");
+
+                if (idprop == null)
+                    throw new InvalidOperationException(String.Format("Unable to determine Id property for type {0}", type));
+
+                idPropCache.Add(type, idprop);
+            }
+
+            return idprop;
+        }
+
+        #endregion
+
+        #region Property Maps
+
+        public Dictionary<string, PropertyInfo> GetPropertyMap(Type type)
+        {
+            Dictionary<string, PropertyInfo> propMap = null;
+
+            var propMapCache = _propertyMaps.Value;
+
+            lock (propMapCache)
+            {
+                if (propMapCache.TryGetValue(type, out propMap)) return propMap;
+
+                propMap = new Dictionary<string, PropertyInfo>();
+                PropertyInfo[] props = type.GetProperties();
+                foreach (PropertyInfo prop in props)
+                {
+                    propMap[JsonApiFormatter.FormatPropertyName(prop.Name)] = prop;
+                }
+
+                propMapCache.Add(type, propMap);
+            }
+
+            return propMap;
+        }
+
+        #endregion
+
+        //TODO: This has been "moved" here so we can cache the results and improve performance...but
+        // it raises the question of whether the various methods called within here should belong
+        // to JsonApiFormatter at all...should they move here also? Should the IPluralizationService
+        // instance belong to ModelManager instead?
+        internal string GetJsonKeyForType(Type type, IPluralizationService pluralizationService)
+        {
+            string key = null;
+
+            var keyCache = _jsonKeysForType.Value;
+
+            lock (keyCache)
+            {
+                if (keyCache.TryGetValue(type, out key)) return key;
+
+                if (JsonApiFormatter.IsMany(type))
+                    type = JsonApiFormatter.GetSingleType(type);
+
+                var attrs = type.CustomAttributes.Where(x => x.AttributeType == typeof(Newtonsoft.Json.JsonObjectAttribute)).ToList();
+
+                string title = type.Name;
+                if (attrs.Any())
+                {
+                    var titles = attrs.First().NamedArguments.Where(arg => arg.MemberName == "Title")
+                        .Select(arg => arg.TypedValue.Value.ToString()).ToList();
+                    if (titles.Any()) title = titles.First();
+                }
+
+                key = JsonApiFormatter.FormatPropertyName(pluralizationService.Pluralize(title));
+
+                keyCache.Add(type, key);
+            }
+
+            return key;
+        }
+    }
+}
```

Comment:
Hmmmmm... Tricky, because `GetGenericArguments`, `GetInterfaces`, and `GetElementType` are probably expensive functions, and we would therefore also benefit from having the result of `IsMany` and `GetSingleType` cached as well, which we can't do if it's an extension on `Type`. I'm going to move them to `IModelManager` as well for now, to get rid of the circular dependency, but I'll probably change their names. We can pretty easily move them again later if we decide to.